### PR TITLE
Remove federation

### PR DESCRIPTION
**Pull request by marun**
This PR removes the federation codebase and associated tooling from the tree.

The first commit just removes the `federation` path and should be uncontroversial.  The second commit removes references and associated tooling and suggests careful review.

Requirements for merge:

- [x] Bazel jobs no longer hard-code federation as a target ([test infra #4983](https://github.com/kubernetes/test-infra/pull/4983))
- [x] `federation-e2e` jobs are not run by default for k/k

**Release note**:

```release-note
Development of Kubernetes Federation has moved to github.com/kubernetes/federation.  This move out of tree also means that Federation will begin releasing separately from Kubernetes.  The impact of this is Federation-specific behavior will no longer be included in kubectl, kubefed will no longer be released as part of Kubernetes, and the Federation servers will no longer be included in the hyperkube binary and image.
```

cc: @kubernetes/sig-multicluster-pr-reviews @kubernetes/sig-testing-pr-reviews

### CODE REVIEW

**Comment by marun** [comment]
The bazel builds will fail so long as the job definitions [refer to the now-missing federation path](https://github.com/kubernetes/test-infra/blob/master/prow/config.yaml#L267).  This suggests defining build targets in the repo rather the job definition.

cc: @kubernetes/sig-testing-misc

**Comment by marun** [comment]
@kubernetes/sig-cli-misc Note the removal of `Cluster` support from `kubectl`.  Support for federation-specific types going forward will be via the discovery client and/or plugin.

**Comment by shashidharatd** [comment]
great work @marun . This looks good to me and finally federation code will be moved out-of-core.
we can just wait for the in-flight federation PR https://github.com/kubernetes/kubernetes/pull/51502 to get merged.

**Comment by shashidharatd** [comment]
Also can you add a release note. maybe its worth mentioning that federation release artifacts won't be part kubernetes release anymore. 
Thanks !

**Comment by mengqiy** [comment]
kubectl bits look good.

This will help reducing the complexity of moving kubectl out of core. Thanks!
cc @monopole

**Comment by jdumars** [comment]
@shashidharatd there's a [discussion on this in flight](https://kubernetes.slack.com/archives/C2C40FMNF/p1509046828000106) in sig-release slack channel - we're in agreement that some final release note should be posted as things move out of tree.

File: build/visible_to/BUILD [hunk 1]
```diff
@@ -93,7 +81,6 @@ package_group(
 package_group(
     name = "pkg_kubectl_CONSUMERS_BAD",
     includes = [
-        ":FEDERATION_BAD",
```
Comment (monopole) [inline build/visible_to/BUILD]: fantastic!

**Comment by marun** [comment]
rebased

**Comment by thockin** [comment]
I nominate this as PR of the quarter :)

**Comment by bacongobbler** [comment]
I've always felt like federation was outside of the scope of core and should've been developed separately from the get go. Well done folks on successfully pulling out WIP features from core to be developed separately and championed by the community. Seconding and nominating this as PR of the year. :)

**Pull request merged**